src/wallet/spend.cpp

```diff
@@ -607,12 +607,15 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& a
         Shuffle(available_coins.other.begin(), available_coins.other.end(), coin_selection_params.rng_fast);
     }
 
+    SelectionResult preselected(preset_inputs.GetSelectionAmount(), SelectionAlgorithm::MANUAL);
+    preselected.AddInput(preset_inputs);
+
     // Coin Selection attempts to select inputs from a pool of eligible UTXOs to fund the
     // transaction at a target feerate. If an attempt fails, more attempts may be made using a more
     // permissive CoinEligibilityFilter.
     std::optional<SelectionResult> res = [&] {
         // Pre-selected inputs already cover the target amount.
-        if (value_to_select <= 0) return std::make_optional(SelectionResult(nTargetValue, SelectionAlgorithm::MANUAL));
+        if (value_to_select <= 0) return std::make_optional(SelectionResult(value_to_select, SelectionAlgorithm::MANUAL));
```

Comment:
In “wallet: account for preselected inputs in target” (e3210a722542a9cb5f7e4be72470dbe488c281fd):

Sorry, if I'm being dense, but how was “deducting the effective values of the preseleted inputs from `m_target`” leading to a different result than “creating a separate SelectionResult with the target set to the effective values of the pre-selected inputs and then starting the selection targeting the delta between that and the search target”?

It sounds to me that in both cases that would be `m_target - effective_value_of_preselection` under the hood? I assume there is a subtle difference here, but it's not obvious to me from the description. Is the difference that once we use `m_target` and “search target” is now composed differently? Please feel free to ignore if this has been covered elsewhere, and I just missed it.

Reply:
The newly added `Merge` function will add the `m_target`s of the two `SelectionResults`. In the event that `value_to_select <= 0`, using `nTargetValue` in this `SelectionResult will add `nTargetValue` to the `m_target` stored in the preset inputs `SelectionResult`. This is incorrect because the target was already met and we're actually just adding it twice.

Also note how all of the other `SelectionResult`s returned by this function all use `value_to_select`.

Reply:
Answering my own question after talking to S3RK out of band:

The correct `m_target` is needed to calculate the change in advance. Previously, the preselection approach was reducing `m_target` to reflect that some funds had already been picked, and now storing them in a SelectionResult and combining the two before calculating the change will be operating on basis of the correct `m_target` instead.

Reply:
Ah thanks, @achow101, the case with the negative target makes it even clearer how this is better.

Reply:
Small extra add to the topic in case you haven't seen it; the pre-set inputs workflow (this stuff) is being decoupled from the coin selection process in #25685. Which should make the flow clearer and less error-prone.